crates/net/src/eventsource/futures.rs

```diff
@@ -0,0 +1,245 @@
+//!
+//! # Example
+//!
+//! ```rust
+//! use gloo_net::eventsource::futures::EventSource;
+//! use wasm_bindgen_futures::spawn_local;
+//! use futures::{stream, StreamExt};
+//!
+//! # macro_rules! console_log {
+//! #    ($($expr:expr),*) => {{}};
+//! # }
+//! # fn no_run() {
+//! let mut es = EventSource::new("http://api.example.com/ssedemo.php").unwrap();
+//! let stream_1 = es.subscribe("some-event-type").unwrap();
+//! let stream_2 = es.subscribe("another-event-type").unwrap();
+//!
+//! spawn_local(async move {
+//!     let mut all_streams = stream::select(stream_1, stream_2);
+//!     while let Some(Ok((event_type, msg))) = all_streams.next().await {
+//!         console_log!(format!("1. {}: {:?}", event_type, msg))
+//!     }
+//!     console_log!("EventSource Closed");
+//! })
+//! # }
+//! ```
+use crate::eventsource::{EventSourceError, State};
+use crate::js_to_js_error;
+use futures_channel::mpsc;
+use futures_core::{ready, Stream};
+use gloo_utils::errors::JsError;
+use pin_project::{pin_project, pinned_drop};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use web_sys::MessageEvent;
+
+/// Wrapper around browser's EventSource API. Dropping
+/// this will close the underlying event source.
+#[allow(missing_debug_implementations)]
+#[derive(Clone)]
+pub struct EventSource {
+    es: web_sys::EventSource,
+}
+
+/// Wrapper around browser's EventSource API.
+#[allow(missing_debug_implementations)]
+#[pin_project(PinnedDrop)]
+pub struct EventSourceSubscription {
+    #[allow(clippy::type_complexity)]
+    error_callback: Closure<dyn FnMut(web_sys::Event)>,
+    es: web_sys::EventSource,
+    event_type: String,
+    message_callback: Closure<dyn FnMut(MessageEvent)>,
+    #[pin]
+    message_receiver: mpsc::UnboundedReceiver<StreamMessage>,
+}
+
+impl EventSource {
+    /// Establish an EventSource.
+    ///
+    /// This function may error in the following cases:
+    /// - The connection url is invalid
+    ///
+    /// The error returned is [`JsError`]. See the
+    /// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/EventSource/EventSource#exceptions_thrown)
+    /// to learn more.
+    pub fn new(url: &str) -> Result<Self, JsError> {
+        let es = web_sys::EventSource::new(url).map_err(js_to_js_error)?;
+
+        Ok(Self { es })
+    }
+
+    /// Subscribes to listening for a specific type of event.
+    ///
+    /// All events for this type are streamed back given the subscription
+    /// returned.
+    ///
+    /// The event type of "message" is a special case, as it will capture
+    /// events without an event field as well as events that have the
+    /// specific type `event: message`. It will not trigger on any
+    /// other event type.
+    pub fn subscribe(&mut self, event_type: &str) -> Result<EventSourceSubscription, JsError> {
```

Comment:
I really like this change (returning a subscription). Nicer API and nicer implementation. Great feedback from @hamza1311 .